### DISCUSSION TITLE
How to set speed depending on button pressed

### DISCUSSION THREAD

**Discussion by rcodecom** [pybricks/support#783] [Q&A] [answered]
Hi, I am looking at the code on this page for the 42109 rally car: 

https://bit.ly/3UlQozr

Everything works fine, but I would like to use the red left button to set the speed to 50%, and use the red right button to set the speed to 100%. Rather than having the speed change **while** the buttons are pressed, I would like just a press of the button to change the speed to the new value where it would remain until the other button is pressed.

How can I do this?

**Comment by dlech** [comment, marked as answer]
Something like this?

```python
    if Button.LEFT in pressed:
        drive_speed = top_speed // 2
    if Button.RIGHT in pressed:
        drive_speed = top_speed
```

**Reply by rcodecom** [reply]
Correct me if I'm wrong but that is what the code in the link does already for adjusting the speed. The issue I have is that the button must be held for the new speed to apply. I want to press and release and have the new speed applied.

**Reply by dlech** [reply]
You are right. I missed the line with with the unconditional `drive_speed = 0`.

If the left button changes the speed to 50% and the right button changes it to 100%, how would you change the speed back to 0?

**Reply by rcodecom** [reply]
All I want to change is the top_speed variable and sustain that without having to keep the button pressed. The plus and minus button functionality remains unchanged to increase and reduce speed.

**Reply by dlech** [reply]
So this then?

```python
    # Top speed controls
    if Button.LEFT in pressed:
        top_speed = 50
    if Button.RIGHT in pressed:
        top_speed = 100
```

**Reply by dlech** [reply]
You will also need to set an initial `top_speed = 100` before the `while True:`.

**Reply by rcodecom** [reply]
Isn't that what the existing code already does? The speed won't maintain itself at 50 for example, unless you keep pressing the left button, which gets uncomfortable quick.

**Reply by dlech** [reply]
Not if we replace the code below with the code I suggested.

```python
    # Top speed controls
    top_speed = 50 # <-- delete this line!
    if Button.LEFT in pressed:
        top_speed = 100  
```

The existing code sets `top_speed` unconditionally in the while loop which is why you have to keep holding the left button to make a change. But if we make it conditional on a button press, then you don't have to keep holding the button since nothing will change the value any more when you release the button.

**Reply by rcodecom** [reply]
Totally right, I missed that. Works beautifully now, thank you.

**Reply by rcodecom** [reply]
As a sidenote, I did have some problems updating the code on the hub after changing some of it as indicated. I had to reset the hub and install the bootloader again, then download the new code for it to be applied. Maybe a bug of sorts with the current version of pybricks.

**Reply by laurensvalk** [reply]
Did you use https://code.pybricks.com/ or https://beta.pybricks.com/ when this issue happened?

**Reply by rcodecom** [reply]
https://code.pybricks.com/

**Reply by laurensvalk** [reply]
For issues with updating code, I suggest trying it out with https://beta.pybricks.com/

We have improved program storage a lot. Resetting/updating is no longer required to save a program. Just run it and it will be saved.

I understand you're experiencing issues with the beta in terms of Bluetooth, however. We're looking into it.

Thanks for taking the time to report your issue here, we really appreciate it :)
